tests/TestCase/ORM/QueryTest.php

```diff
@@ -2221,6 +2221,64 @@ public function testFormatResults()
         $this->assertSame([$callback2, $callback1], $query->getResultFormatters());
     }
 
+    /**
+     * Tests that results formatters do receive the query object.
+     *
+     * @return void
+     */
+    public function testResultFormatterReceivesTheQueryObject()
+    {
+        $resultFormatterQuery = null;
+
+        $query = $this->getTableLocator()->get('authors')
+            ->find()
+            ->formatResults(function ($results, $query) use (&$resultFormatterQuery) {
+                $resultFormatterQuery = $query;
+
+                return $results;
+            });
+        $query->firstOrFail();
+
+        $this->assertSame($query, $resultFormatterQuery);
+    }
+
+    /**
+     * Tests that results formatters for queries of joined associations
+     * do receive the source query, not the association target query.
+     *
+     * @return void
+     */
+    public function testResultFormatterReceivesTheSourceQueryForJoinedAssociations()
+    {
+        $articles = $this->getTableLocator()->get('Articles');
+        $authors = $articles->belongsTo('Authors');
+
+        $resultFormatterTargetQuery = null;
+        $resultFormatterSourceQuery = null;
+
+        $authors->getEventManager()->on(
+            'Model.beforeFind',
+            function ($event, Query $targetQuery) use (&$resultFormatterTargetQuery, &$resultFormatterSourceQuery) {
+                $resultFormatterTargetQuery = $targetQuery;
+
+                $targetQuery->formatResults(function ($results, $query) use (&$resultFormatterSourceQuery) {
```

Comment:
Won't this behave differently for different association types?

Reply:
I'm not sure what exactly you're pointing to, but it would behave differently in so far as that for non-joined associations, the formatter would receive the association target query instead instead of source query.

I can add a test to show the different behavior.

Reply:
I clicked on the wrong line, but yes that's what I mean.

I think that might be confusing, but something we could possibly fix with clear documentation.

Reply:
Yeah, if accepted, I'll add some explanation in the `formatResults()` docblock once the code is finished.

Reply:
Wait a minute, I remember this section in the docs very differently: https://book.cakephp.org/4/en/orm/query-builder.html#adding-calculated-fields

I thought it would be a good opportunity to expand on it, but it actually already says that the second argument would be a query instance 🙃 I must be mixing things up, I remember this section as a short only recently ago.